mpt_trie/src/trie_ops.rs

```diff
@@ -513,58 +548,61 @@ fn delete_intern<N: PartialTrie>(
                 .nibbles_are_identical_up_to_smallest_count(&curr_k)
                 .then(|| {
                     curr_k.truncate_n_nibbles_front_mut(ext_nibbles.count);
-                    delete_intern(child, curr_k).map(|(updated_child, value_deleted)| {
-                        let updated_node = collapse_ext_node_if_needed(ext_nibbles, &updated_child);
-                        (updated_node, value_deleted)
+
+                    delete_intern(child, curr_k).and_then(|res| {
+                        res.map_or(Ok(None), |(updated_child, value_deleted)| {
+                            let updated_node =
+                                collapse_ext_node_if_needed(ext_nibbles, &updated_child)?;
+                            Ok(Some((updated_node, value_deleted)))
+                        })
                     })
                 })
-                .flatten()
+                .unwrap_or(Ok(None))
         }
         Node::Leaf { nibbles, value } => {
             trace!("Delete traversed Leaf (nibbles: {:?})", nibbles);
-            (*nibbles == curr_k).then(|| {
-                trace!("Deleting leaf ({:x})", nibbles);
-                (Node::Empty.into(), value.clone())
-            })
+            (*nibbles == curr_k)
+                .then(|| {
+                    trace!("Deleting leaf ({:x})", nibbles);
+                    Ok(Some((Node::Empty.into(), value.clone())))
+                })
+                .unwrap_or(Ok(None))
```

Comment:
I _think_ we can wrap the block in an `Ok()` here as well? Less sure for this case.